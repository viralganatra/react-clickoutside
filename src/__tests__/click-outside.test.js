import React from 'react';
import ClickOutside from '../click-outside';
import EventName from '../events';

jest.mock('../events', () => 'mousedown');

describe('Click Outside Render', () => {
    const globalAddEvent = global.addEventListener;
    const globalRemoveEvent = global.removeEventListener;

    let eventMap;
    let wrapper;

    beforeEach(() => {
        eventMap = {};

        global.addEventListener = jest.fn((event, cb) => {
            event && (eventMap[event] = cb);
        });

        global.removeEventListener = jest.fn((event) => {
            event && (delete eventMap[event]);
        });

        wrapper = mount(
            <div>
                <p className="title">Title</p>
                <ClickOutside>
                    {({ hasClickedOutside }) => {
                        const text = hasClickedOutside ?
                            'Clicked outside' :
                            'Has not clicked outside';

                        return <div className="container">{text}</div>;
                    }}
                </ClickOutside>
            </div>,
        );
    });

    afterEach(() => {
        global.addEventListener = globalAddEvent;
        global.removeEventListener = globalRemoveEvent;
    });

    it('should detect the click event outside of the component', () => {
        eventMap.mousedown({ target: wrapper.find('.title').instance() });

        wrapper.update();

        expect(wrapper).toMatchSnapshot();
    });

    it('should detect the click event inside of the component', () => {
        eventMap.mousedown({ target: wrapper.find('.container').instance() });

        wrapper.update();

        expect(wrapper).toMatchSnapshot();
    });

    it('should unregister the event listener when the component unmounts', () => {
        wrapper.unmount();

        expect(global.removeEventListener.mock.calls[3][0]).toBe(EventName);
    });

    it('should add any custom properties to the container div', () => {
        const wrapperProps = mount(
            <ClickOutside className="foo">
                {() => {}}
            </ClickOutside>,
        );

        expect(wrapperProps).toMatchSnapshot();
    });
});
