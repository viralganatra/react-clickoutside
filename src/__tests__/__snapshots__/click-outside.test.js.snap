// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Click Outside Render should add any custom properties to the container div 1`] = `
<ClickOutside
  className="foo"
>
  <div
    className="foo"
  />
</ClickOutside>
`;

exports[`Click Outside Render should detect the click event inside of the component 1`] = `
<div>
  <p
    className="title"
  >
    Title
  </p>
  <ClickOutside>
    <div>
      <div
        className="container"
      >
        Has not clicked outside
      </div>
    </div>
  </ClickOutside>
</div>
`;

exports[`Click Outside Render should detect the click event outside of the component 1`] = `
<div>
  <p
    className="title"
  >
    Title
  </p>
  <ClickOutside>
    <div>
      <div
        className="container"
      >
        Clicked outside
      </div>
    </div>
  </ClickOutside>
</div>
`;
